gokart/gcs_obj_metadata_client.py

```diff
@@ -101,23 +107,49 @@ def _get_patched_obj_metadata(
         # Instead, users are expected to search using the labels they provided.
         # Therefore, in the event of a key conflict, the value registered by the user-provided labels will take precedence.
-        _merged_labels = GCSObjectMetadataClient._merge_custom_labels_and_task_params_labels(normalized_task_params_labels, normalized_custom_labels)
+        normalized_labels = (
+            [normalized_custom_labels, normalized_task_params_labels]
+            if not required_task_outputs
+            else [
+                normalized_custom_labels,
+                normalized_task_params_labels,
+                {'__required_task_outputs': json.dumps(GCSObjectMetadataClient._get_serialized_string(required_task_outputs))},
+            ]
+        )
+        _merged_labels = GCSObjectMetadataClient._merge_custom_labels_and_task_params_labels(normalized_labels)
         return GCSObjectMetadataClient._adjust_gcs_metadata_limit_size(dict(metadata) | _merged_labels)
 
+    @staticmethod
+    def _get_serialized_string(required_task_outputs: FlattenableItems[RequiredTaskOutput]) -> FlattenableItems[str]:
+        def _iterable_flatten(nested_list: Iterable) -> list[str]:
+            flattened_list: list[str] = []
+            for item in nested_list:
+                if isinstance(item, Iterable):
+                    flattened_list.extend(_iterable_flatten(item))
+                else:
+                    flattened_list.append(item)
+            return flattened_list
+
+        if isinstance(required_task_outputs, dict):
+            return {k: GCSObjectMetadataClient._get_serialized_string(v) for k, v in required_task_outputs.items()}
+        if isinstance(required_task_outputs, Iterable):
+            return _iterable_flatten([GCSObjectMetadataClient._get_serialized_string(ro) for ro in required_task_outputs])
+        return [required_task_outputs.serialize()]
+
     @staticmethod
     def _merge_custom_labels_and_task_params_labels(
-        normalized_task_params: dict[str, str],
-        normalized_custom_labels: dict[str, Any],
+        normalized_labels_list: list[dict[str, Any]],
     ) -> dict[str, str]:
-        merged_labels = copy.deepcopy(normalized_custom_labels)
-        for label_name, label_value in normalized_task_params.items():
-            if len(label_value) == 0:
-                logger.warning(f'value of label_name={label_name} is empty. So skip to add as a metadata.')
-                continue
-            if label_name in merged_labels.keys():
-                logger.warning(f'label_name={label_name} is already seen. So skip to add as a metadata.')
-                continue
-            merged_labels[label_name] = label_value
+        merged_labels: dict[str, str] = {}
+        for normalized_label in normalized_labels_list[:]:
+            for label_name, label_value in normalized_label.items():
+                if len(label_value) == 0:
```

Comment:
[MUST] This code may fail, since it seems to assume that `label_value ` is str.

I prefer checking if it is str, and then check the length as,

```
isinstance(label_value, str) and len(label_value)==0
```

Reply:
@TlexCypher 
Then maybe the input `normalized_labels_list: list[dict[str, Any]]` should be `normalized_labels_list: list[dict[str, str]]` ?

Reply:
I fixed here  3a60c45